apps/events/viewsets.py

```diff
@@ -101,17 +111,23 @@ def get_template_names(self):
                 return ["events/detail.html"]
 
     def get_queryset(self):
-        if self.kwargs.get("pk"):
-            self.queryset = Event.objects.filter(id=self.kwargs["pk"])
+        if self.kwargs.get("event_id"):
+            self.queryset = Event.objects.filter(id=self.kwargs["event_id"])
+            return self.queryset.all().annotate(participants_number=Count("participants"))
+        if self.kwargs.get("token"):
+            self.queryset = Event.objects.filter(private_url=self.kwargs["token"])
+            return self.queryset.all().annotate(participants_number=Count("participants"))
         else:
             if self.request.user.id:
                 self.queryset = self.model.objects.filter(
-                    Q(is_visible=True) & Q(is_finished=False)
-                    | Q(participants__in=[self.request.user.id]) & Q(is_finished=False)
+                    Q(is_visible=True) & Q(is_finished=False) & Q(type="open") |
+                    Q(participants__in=[self.request.user.id]) & Q(type="private") |
+                    Q(created_by=self.request.user) & Q(type="private")
```

Comment:
For all these filter we should have an option toggled Q(is_visible=True) & Q(is_finished=False), that is main criteria for showing event on the main page, if some of these flags is not set, it should not be visible on the main page.